bpf/process/types/basic.h

```diff
@@ -560,10 +560,13 @@ filter_char_buf(struct selector_arg_filter *filter, char *args)
 		asm volatile("%[j] &= 0xff;\n" ::[j] "+r"(j)
 			     :);
 		length = *(__u32 *)&value[j];
-		asm volatile("%[length] &= 0x3f;\n" ::[length] "+r"(length)
+		asm volatile("%[length] &= 0x3f;\n" ::[length] "+r"(
+				     length)
 			     :);
-		v = (int)value[j];
-		a = (int)args[0];
+		// bytes of the matchArgs value
+		v = *(int *)&value[j];
+		// bytes of the char_buf
+		a = ((int *)args)[0];
```

Comment:
I think a needs to be `a = *(int *)&args[value_off - 4];` which is the length that was actually read from the buffer. In 99% of the cases, it will be the same as (int *)&args[0] so it's not easy to test this.

Reply:
i think `a = *(int *)&args[value_off - 4];` is imprecise. for char_buf type, `a = *(int *)&args[value_off - 4]` is the length that may be truncated，but `((int *)args)[0]` is the real length.

Reply:
I agree. The question is what do we need to use for the comparison the real or the truncated length.

My thinking was that we need the truncated length because that's the actual number of bytes we can compare. If we compare using the real length, we might be comparing garbage.

Reply:
I agree. Only truncated length seems to make sense. If the buffer has been truncated then an exact match will not match anyway, and for partial (prefix) matches, you can only match what's actually in the buffer. In all these cases, the real length is only useful if you have a test/comparison on original length only, which I doubt we have (is this buffer >=X bytes long?).

Reply:
mabye you are right，char_buf will only be truncated when the original length is greater than 4096